### ISSUE TITLE
[WebProfiler] Expressions could get a better display in the panel

### ISSUE THREAD

**Issue by stof** [symfony/symfony#24127]
Currently, voting on an expression displays a panel like that:

![expression_display](https://user-images.githubusercontent.com/439401/30164953-81b1349c-93de-11e7-9934-7b3420ab94b8.png)

(This is a SerializedParsedExpression rather than just an Expression because it is the vote for an expression used in the access_control).

We could improve this display when encountering an Expression object by displaying its string representation instead (which will be the expression itself).

**Comment by javiereguiluz** [comment]
I agree. The `attributes` table cell is displayed with:

```twig
<td>
    {% if decision.attributes|length == 1 %}
        {{ decision.attributes|first }}
    {% else %}
        {{ profiler_dump(decision.attributes) }}
    {% endif %}
</td>
```

So this must be done in `profiler_dump()`. @nicolas-grekas do you think this is feasible and can be implemented without breaking other features? Thanks!

**Comment by stof** [comment]
this does not look like a VarDumper output in my screenshot

**Comment by stof** [comment]
OK, the issue is that `{{ decision.attributes|first }}` casts the first attribute to a string. But the attribute is not a SerializedParsedExpression anymore here, but a VarDumper Data object.

Btw, when I dump the internals of this object, the `expression` property seems to contain lots of crap instead of a string value, so it looks like the VarDumper is buggy there (using 3.3.8).

I think we would have to cast the Expression object to string when collecting the access decision log instead.

**Comment by stof** [comment]
Another solution I found:

```twig
{% if decision.attributes|length == 1 %}
    {% set attribute = decision.attributes|first %}
    {% if attribute.type in ['Symfony\\Component\\ExpressionLanguage\\SerializedParsedExpression', 'Symfony\\Component\\ExpressionLanguage\\ParsedExpression', 'Symfony\\Component\\ExpressionLanguage\\Expression'] %}
        {% set attribute = attribute.expression %}
    {% endif %}
    {{ attribute }}
{% else %}
    {{ profiler_dump(decision.attributes) }}
{% endif %}
```

**Comment by nicolas-grekas** [comment]
@stof can you post a reproducer for the "crap" issue? That may help solve other related ones?

**Comment by stof** [comment]
@nicolas-grekas here is the output of doing a `dump(decision.attributes)` in this place in the code of the template (the last line in the screen is the existing `{{ decision.attributes|first }}`):

<details>
<summary>Internal data dump</summary>

![image](https://user-images.githubusercontent.com/439401/36479747-11bd62a0-170a-11e8-882a-b5b64007ad70.png)
</details>

And here is the output when doing `profiler_dump(decision.attributes)`:
![image](https://user-images.githubusercontent.com/439401/36479815-5b7a2be4-170a-11e8-8c63-c4b697363b48.png)

But looking more in details, this seems related to the way Data works (it always contains the whole data, and a pointer to the right place. So it may not be crap, just misunderstanding from my part.

I found a way to make the code work, with a small change to my code:

```twig
{% if decision.attributes|length == 1 %}
    {% set attribute = decision.attributes|first %}
    {% if attribute.type in [
        'Symfony\\Component\\ExpressionLanguage\\SerializedParsedExpression',
        'Symfony\\Component\\ExpressionLanguage\\ParsedExpression',
        'Symfony\\Component\\ExpressionLanguage\\Expression'
    ] %}
        {% set attribute = attribute["\x00*\x00expression"] %}
    {% endif %}
    {{ attribute }}
{% else %}
    {{ profiler_dump(decision.attributes) }}
{% endif %}
```

My mistake was to try to access `expression`, while the Data object had `\x00*\x00expression` as key due to being a protected property.

It then looks like this:
![image](https://user-images.githubusercontent.com/439401/36480883-39ec67a4-170e-11e8-9ca6-fc1ef05bdb3e.png)

the drawback of this proposal is that it does not make it clear that it voted on an expression (depending on how it looks).

Here is an alternate proposal (with the advantage of providing a readable output even when voting on other kind of objects as attribute):

```twig
{% if decision.attributes|length == 1 %}
    {% set attribute = decision.attributes|first %}
    {% if attribute.type == 'string' %}
        {{ attribute }}
    {% else %}
         {{ profiler_dump(attribute) }}
    {% endif %}
    {{ attribute }}
{% else %}
    {{ profiler_dump(decision.attributes) }}
{% endif %}
```
![image](https://user-images.githubusercontent.com/439401/36481015-bef7058a-170e-11e8-8335-2c0836db44d6.png)
![capture d ecran 2018-02-21 a 13 55 00](https://user-images.githubusercontent.com/439401/36481050-db4ea440-170e-11e8-8a4e-4d6568ebc0f1.png)

This could then be improved with a custom caster for ParsedExpression hiding the `nodes` property.

A third alternative is to handle expressions in a special way at collecting time (casting them to string), but this feels weird.

I would go for the second alternative (handling any non-string attribute properly in the template instead of hardcoding a special case for expressions).
What do you think @symfony/deciders ?

**Comment by lyrixx** [comment]
I added a way to dump back to "string" an expression in Symfony few years ago
cf https://github.com/symfony/symfony/pull/19722 and  https://github.com/symfony/symfony/pull/19013
We could leverage this dumper to dump a string representation.

I also added in blackfire (private repository) a way to dump the expression with associated variables 
[example](https://blackfire.io/docs/gitlist-twig-symfony-assertions.png).

If @fabpot agree, we could "move" this code to the Symfony profiler.
